capa/features/extractors/binexport2/arch/intel/insn.py

```diff
@@ -0,0 +1,249 @@
+            if be2.expression[expression_index].type == BinExport2.Expression.DEREFERENCE:
+                is_dereference = True
+                break
+
+        if not is_dereference:
+            continue
+
+        phrase_info: Optional[OperandPhraseInfo] = get_operand_phrase_info(be2, operand)
+        if not phrase_info:
+            continue
+
+        if phrase_info.displacement:
+            if phrase_info.base and phrase_info.base.symbol.lower().endswith(("bp", "sp")):
+                # skips things like:
+                # 00401068 MOV dword ptr [EBP + local_8],EAX
+                continue
+
+            value = mask_immediate(fhi.arch, phrase_info.displacement.immediate)
+            if not is_address_mapped(be2, value):
+                value = capa.features.extractors.binexport2.helpers.twos_complement(fhi.arch, value, 32)
+
+                yield Offset(value), ih.address
+                yield OperandOffset(i, value), ih.address
+
+                if mnemonic == "lea" and i == 1:
+                    if phrase_info.base and not any((phrase_info.scale, phrase_info.index)):
+                        yield Number(value), ih.address
+                        yield OperandNumber(i, value), ih.address
+
+        elif phrase_info.base and not any((phrase_info.index, phrase_info.scale)):
+            # like:
+            # 00401062 MOVZX EAX,word ptr [EDI]
+            yield Offset(0), ih.address
+            yield OperandOffset(i, 0), ih.address
+
+
+def is_security_cookie(
+    fhi: FunctionContext,
+    bbi: BasicBlockContext,
+    instruction: BinExport2.Instruction,
+) -> bool:
+    """
+    check if an instruction is related to security cookie checks.
+    """
+    be2: BinExport2 = fhi.ctx.be2
+
+    # security cookie check should use SP or BP
+    op1: BinExport2.Operand = be2.operand[instruction.operand_index[1]]
+    op1_exprs: List[BinExport2.Expression] = [be2.expression[expr_i] for expr_i in op1.expression_index]
+    if all(expr.symbol.lower() not in ("bp", "esp", "ebp", "rbp", "rsp") for expr in op1_exprs):
+        return False
+
+    # check_nzxor_security_cookie_delta
+    # if insn falls at the start of first entry block of the parent function.
+    flow_graph: BinExport2.FlowGraph = be2.flow_graph[fhi.flow_graph_index]
+    basic_block_index: int = bbi.basic_block_index
+    bb: BinExport2.BasicBlock = be2.basic_block[basic_block_index]
+    if flow_graph.entry_basic_block_index == basic_block_index:
+        first_addr: int = min((be2.instruction[ir.begin_index].address for ir in bb.instruction_index))
+        if instruction.address < first_addr + SECURITY_COOKIE_BYTES_DELTA:
+            return True
+    # or insn falls at the end before return in a terminal basic block.
+    if basic_block_index not in (e.source_basic_block_index for e in flow_graph.edge):
+        last_addr: int = max((be2.instruction[ir.end_index - 1].address for ir in bb.instruction_index))
```

Comment:
Can we rely on `instruction.address` to be present for these items? I recall that this field is not set for all instructions, and sometimes must be recovered using the sizes of preceding instructions.

Reply:
we can use `idx.instruction_address_by_index` perhaps

Reply:
Correct, BinExport2 does not store the address of every instruction, we should update this code to use `idx.instruction_address_by_index` instead.

Reply:
- [ ] use `idx.instruction_address_by_index` when accessing instruction address

Reply:
i'll have a PR for this shortly